vars/unitTestPost.groovy

```diff
@@ -26,90 +26,115 @@
    * config['valgrind_stash']      Name to stash valgrind artifacts
    *                               Required if more than one stage is
    *                               creating valgrind reports.
-   *
    */
 
-def call(Map config = [:]) {
-
-  String always_script = config.get('always_script',
-                                    'ci/unit/test_post_always.sh')
-  sh label: 'Job Cleanup',
-     script: always_script
-
-  Map stage_info = parseStageInfo(config)
+// groovylint-disable DuplicateNumberLiteral, DuplicateStringLiteral,
+// groovylint-disable VariableName
 
-  if (config['testResults'] != 'None' ) {
-    double health_scale = 1.0
-    if (config['ignore_failure']) {
-      health_scale = 0.0
-    }
+// groovylint-disable-next-line MethodSize
+void call(Map config = [:]) {
+    String always_script = config.get('always_script',
+                                      'ci/unit/test_post_always.sh')
+    sh label: 'Job Cleanup',
+       script: always_script
 
-    def cb_result = currentBuild.result
-    junit testResults: config.get('testResults', 'test_results/*.xml'),
-          healthScaleFactor: health_scale
+    Map stage_info = parseStageInfo(config)
 
-    if (cb_result != currentBuild.result) {
-      println "The junit plugin changed result to ${currentBuild.result}."
+    if (config['testResults'] != 'None' ) {
+        // groovylint-disable-next-line NoDouble
+        double health_scale = 1.0
+        if (config['ignore_failure']) {
+            health_scale = 0.0
+        }
+        String cb_result = currentBuild.result
+        junit testResults: config.get('testResults', 'test_results/*.xml'),
+              healthScaleFactor: health_scale
+        if (cb_result != currentBuild.result) {
+            println "The junit plugin changed result to ${currentBuild.result}."
+        }
     }
-  }
-
-  if(stage_info['with_valgrind']) {
-    String target_dir = "unit_test_memcheck_logs"
-    String src_files = "unit-test-*.memcheck.xml"
-    fileOperations([fileCopyOperation(excludes: '',
+    if (stage_info['with_valgrind']) {
+        String target_dir = 'unit_test_memcheck_logs'
+        String src_files = 'unit-test-*.memcheck.xml'
+        int vgfail = 0
+        int vgerr = 0
+        String rcs = sh label: 'Check for Valgrind errors',
+                 script: "grep -E '<error( |>)' ${src_files} || true",
+                 returnStdout: true
+        if (rcs) {
+            if (config['ignore_failure']) {
+                vgfail = 1
+            } else {
+                vgerr = 1
+            }
```

Comment:
Per the other comments, this seems to unnecessarily contort the meaning of `ignore_failure` for Valgrind stages so that it has a different and inconsistent meaning.  I think we can preserve the meaning and consistency of `ignore_failure` and simply change the caller to not request it for the Valgrind stage if the reporter of SRE-1729 doesn't want failures in that stage ignored.
```suggestion
            vgfail = 1
```

Reply:
With ignore_failure set, a test failure in a stage marks it as unstable, but the pipeline continues, and a test error will stop the pipeline when Junit reports it.

Current behavior is that with ignore_failure clear, a script return of non-zero or a junit test result of "error" will stop the pipeline, but a junit test result of "failure" will not.

The vgfail=1 causes Junit to mark the stage unstable if it is not already unstable.
The vgerr=1 causes Junit to mark the stage as failed if it is not already failed.

With vgfail always = 1 and ignore failure clear, the stage will not stop the pipeline on a valgrind issue.

Will put that in the next push.

Reply:
> With ignore_failure set, a test failure in a stage marks it as unstable, but the pipeline continues, and a test error will stop the pipeline when Junit reports it.

The pipeline may _continue_ after that `UNSTABLE` stage, however https://github.com/daos-stack/pipeline-lib/blob/master/vars/skipStage.groovy#L415 and https://github.com/daos-stack/pipeline-lib/blob/master/vars/skipStage.groovy#L529 are supposed to skip the subsequent stages if any previous stage ends up `UNSTABLE`.  That's the mechanism that should be used here, not forcing an `UNSTABLE` (yellow) stage into `FAILURE` (red).

> Current behavior is that with ignore_failure clear,

By _clear_ you mean `false`, correct?

> a script return of non-zero or a junit test result of "error" will stop the pipeline, but a junit test result of "failure" will not.

Which is why we introduced skipping on `UNSTABLE` into [`skipStage()`](https://github.com/daos-stack/pipeline-lib/blob/master/vars/skipStage.groovy)

> The vgfail=1 causes Junit to mark the stage unstable if it is not already unstable. The vgerr=1 causes Junit to mark the stage as failed if it is not already failed.

Understood.  But we should not be converting `UNSTABLE` info `FAILURE` just to effect pipeline stoppage.  We have `skipStage()` and it's check for `UNSTABLE` (and even a commit pragma to override that) to achieve that.

> With vgfail always = 1 and ignore failure clear, the stage will not stop the pipeline on a valgrind issue.

Again, looking at the bigger picture, I don't think Valgrind errors should ever make the stage `FAILED` (red) under any conditions.  Only failing to actually execute the test in that stage should make it `FAILED` (red).  After-all, a Valgrind error is not because the test failed to execute.  The test executed successfully.  It just reported (Valgrind, in this case) errors.  That is a test failure and the result should be `UNSTABLE` (yellow).

Reply:
The settings for if Valgrind fails or marks a build unstable are currently controlled by the valgrindReportPublish.groovy code, and that step was not written to allow a DAOS PR to modify them.  I would prefer any changes there be done as a different PR than this one.

And yes, by ignore_failures being clear, I mean false.  However the current daos/master and daos/release never return a non-zero result for unit tests that fail.

https://build.hpdd.intel.com/job/daos-stack/job/daos/view/change-requests/job/PR-11595/48/ is with DAOS modified to  induce a memory leak and return a non-zero result for unit tests that fail.  Since DAOS is now returning a non-zero result, the code to check if if there were failures reported in Junit tests is not run.

I will need to run another DAOS test PR to check that code.

Reply:
> The settings for if Valgrind fails or marks a build unstable at the end of the run are currently controlled by the valgrindReportPublish.groovy code, and that step was not written to allow a DAOS PR to modify them. I would prefer any changes there be done as a different PR than this one.

That's fair enough.  Given that it's a defect in the Jenkins plugin (it should be `UNSTABLE` for test failures, not `FAILURE`) this is just another Jenkins bug we are going to have to learn to live with.  That is assuming of course the origin of the XML that we are feeding to the plugin.  What generates that?  Is it valgrind itself?  Maybe it should be distinguishing between `error` and `fail` better?

> Note: The Valgrind Publish step setting the job to failed status does not show up in the Stage View, Blue Ocean, or in the new Pipeline Graph. It only shows up in the console log because we put a check for that in a wrapper to the Valgring Publish step.

Which is exactly why I had requested we generate our own generic JUnit failure to tell the user about it.

> However the current daos/master and daos/release never return a non-zero result for unit tests that fail

If the unit test framework runs successfully (regardless of actual test success/failure) it should return `0`.  It should only return `!0` if it failed to run tests for some reason.  Actual test failures should be reported by JUnit results.